ldap/servers/plugins/sync/sync_persist.c

```diff
@@ -1072,6 +1073,12 @@ sync_send_results(void *arg)
     slapi_ch_free((void **)&strFilter);
     slapi_pblock_set(req->req_pblock, SLAPI_SEARCH_STRFILTER, NULL);
 
+    slapi_pblock_get(req->req_pblock, SLAPI_REQCONTROLS, ctrls);
```

Comment:
Sorry for my late answer. There is something I do not understand here: where this control in req_block comes from ?

We have a client sync_persist request with a control (sync_persist). This control is in the original SLAPI_REQCONTROLS. I initially thought it was erroneously copied in sync_pblock_copy but it looks it is not. So the only step where it could be set is in slapi_send_ldap_search_entry/send_ldap_search_entry/sync_send_results but here it is set in SLAPI_SEARCH_CTRLS not in SLAPI_REQCONTROLS.

In short I fail to find where this control comes from. I think the initial control (sync) should not be copied in req_block because it is only useful to select if it is a refresh/persist and useless in sync_send_results. But anyway I do not see where it is copy
If the coming from the sent results (set in SLAPI_SEARCH_CTRLS), how can it be retrieved in SLAPI_REQCONTROLS

Reply:
> Sorry for my late answer. There is something I do not understand here: where this control in req_block comes from ?

Here is what I think happens.


**Allocation**
When a client binds with ldapv3 we check for ldap request controls: https://github.com/389ds/389-ds-base/blob/56939cc3017d9527a30d4816ed75c654b096c371/ldap/servers/slapd/bind.c#L227

If they exist we allocate memory for the req ctrl, iterate through the ber elements populating the req ctrls: https://github.com/389ds/389-ds-base/blob/56939cc3017d9527a30d4816ed75c654b096c371/ldap/servers/slapd/control.c#L259

Then set the req ctrl in the pblock: https://github.com/389ds/389-ds-base/blob/56939cc3017d9527a30d4816ed75c654b096c371/ldap/servers/slapd/control.c#L357

**Use**
The sync repl plugin registers a pre search function: https://github.com/389ds/389-ds-base/blob/56939cc3017d9527a30d4816ed75c654b096c371/ldap/servers/plugins/sync/sync_init.c#L125

If a client performs a syn repl search in refreshPersist mode then a thread is kicked off to send results back to the client:
https://github.com/389ds/389-ds-base/blob/56939cc3017d9527a30d4816ed75c654b096c371/ldap/servers/plugins/sync/sync_refresh.c#L116

**Leak**
The thread function (sync_send_results) does its magic and terminates. But when the sync repl client closes the connection the req ctrls memory is never freed up. Hence the mem leak highlighted by asan.

Reply:
Do you have the backtrace where the leaking control is allocated ?

I am surprise that we could retrieve in sync SRCH a control allocated during the BIND. I feel like only SRCH control can be stored in the req_pblock.

Reply:
```
Indirect leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c708ba097 in calloc (/lib64/libasan.so.8+0xba097)
    #1 0x7f5c7006ac25 in slapi_ch_calloc (/usr/lib64/dirsrv/libslapd.so.0+0x26ac25)
    #2 0x7f5c7006b7a1 in get_ldapmessage_controls_ext (/usr/lib64/dirsrv/libslapd.so.0+0x26b7a1)
    #3 0x55639e2aadee in do_search ldap/servers/slapd/search.c:349
    #4 0x55639e2813fe in connection_dispatch_operation ldap/servers/slapd/connection.c:677
    #5 0x55639e2813fe in connection_threadmain ldap/servers/slapd/connection.c:1805
    #6 0x7f5c70f39412 in _pt_root (/lib64/libnspr4.so+0x2c412)
    #7 0x7f5c6fa8cdec in start_thread (/lib64/libc.so.6+0x8cdec)
    #8 0x7f5c6fb1236f in clone3 (/lib64/libc.so.6+0x11236f)


Indirect leak of 25 byte(s) in 1 object(s) allocated from:
    #0 0x7f5c708ba6af in __interceptor_malloc (/lib64/libasan.so.8+0xba6af)
    #1 0x7f5c70fc6ed4 in ber_memalloc_x (/lib64/liblber.so.2+0x6ed4)
    #2 0x7f5c70fc6f98 in ber_get_stringbv (/lib64/liblber.so.2+0x6f98)
    #3 0x7f5c70fc734d in ber_get_stringa (/lib64/liblber.so.2+0x734d)
    #4 0x7f5c70fc9820 in ber_scanf (/lib64/liblber.so.2+0x9820)
    #5 0x7f5c7006b813 in get_ldapmessage_controls_ext (/usr/lib64/dirsrv/libslapd.so.0+0x26b813)
    #6 0x55639e2aadee in do_search ldap/servers/slapd/search.c:349
    #7 0x55639e2813fe in connection_dispatch_operation ldap/servers/slapd/connection.c:677
    #8 0x55639e2813fe in connection_threadmain ldap/servers/slapd/connection.c:1805
    #9 0x7f5c70f39412 in _pt_root (/lib64/libnspr4.so+0x2c412)
    #10 0x7f5c6fa8cdec in start_thread (/lib64/libc.so.6+0x8cdec)
    #11 0x7f5c6fb1236f in clone3 (/lib64/libc.so.6+0x11236f)
```

Reply:
Excellent.
so this is the control within the original SRCH req that is set in SLAPI_REQCONTROLS. My understanding it is the sync control (along with the type of sync refresh/persist and the cookie). I think this control should not go into req_pblock that is used to send back result and do check the control. I wonder if the control is not erroneously moved to req_pblock for example during sync_persist_add.

Reply:
I ran the CI test with asan before and after this change, it does remove the memory leak.